Code/GraphMol/Resonance.h

```diff
@@ -29,7 +29,60 @@ typedef std::map<unsigned int, AtomElectrons *> ConjAtomMap;
 typedef std::vector<ConjElectrons *> CEVect;
 typedef std::vector<CEVect2 *> CEVect3;
 typedef std::vector<std::uint8_t> ConjFP;
-typedef boost::unordered_map<std::size_t, ConjElectrons *> CEMap;
+typedef std::unordered_map<std::size_t, ConjElectrons *> CEMap;
+
+/*!
+ * Create a derived class from this abstract base class
+ * and implement the callback() method.
+ * The callback() method is called at each iteration of the
+ * algorithm, and provides a mechanism to monitor or stop
+ * its progress.
+ * To have your callback called, pass an instance of your
+ * derived class to ResonanceMolSupplier::setProgressCallback().
+ */
+class RDKIT_GRAPHMOL_EXPORT ResonanceMolSupplierCallback {
+  friend class ResonanceMolSupplier;
+
+ public:
+  ResonanceMolSupplierCallback() {}
+  virtual ~ResonanceMolSupplierCallback() {}
+  /*! Get the number of conjugated groups this molecule has. */
+  unsigned int getNumConjGrps() const { return d_progress.size(); }
+  /*! Get the maximum number of resonance structures the supplier
+      is allowed to generate.
+      \return The number of structures.
+   */
+  size_t getMaxStructures() const { return d_maxStructs; }
+  /*! Get the number of resonance structures generated so far
+      for a certain conjugated group.
+      \param conjGrpIdx the conjugated group index.
+      \return the number of resonance structures generated so far
+              for this group.
+   */
+  size_t getNumStructures(unsigned int conjGrpIdx) const;
+  /*! Get the number of resonance structures with a different score
+      (i.e., non-degenerate) generated so far for a certain conjugated
+      group.
+      \param conjGrpIdx the conjugated group index.
+      \return the number of non-degenrate resonance structures generated
+              so far for this group.
+   */
+  size_t getNumDiverseStructures(unsigned int conjGrpIdx) const;
+  /*! Pure virtual; this must be implemented in the derived class.
+      \return true if the resonance structure generation should continue;
+              false if the resonance structure generation should stop.
+   */
+  virtual bool callback() const = 0;
```

Comment:
Any particular reason you chose to do this with a `callback()` method instead of using `operator()`?

Reply:
@greglandrum Ouch, it looks like I missed the GitHub notifications and I have only seen this now. No, not really, except it is the same thing I did for MCS... to which your question would apply too! I have noticed that in a similar context you have indeed overridden `operator()` and `__call__()` respectively, which might be more intuitive. I am happy to switch to using the operators here and doing the same in MCS, deprecating the callback function. Please let me know if this sounds good to you.

Reply:
@ptosco I don't think it's strictly necessary, and there's clearly no consistency argument to be made, but using functors somehow seems nicer to me. Since I don't think there's a "right" answer I defer to you the decision (both here and on the MCS code). :-)

If you do decide to change the MCS code, please do deprecate the existing method name (with an entry in the release notes) and we can actually remove the `callback()` method itself in the 2021.03 release

Reply:
@greglandrum OK, will do the changes both here and in the MCS code, and add tests, then I’ll revise this PR. Thanks!

Reply:
@ptosco just to confirm: I should hold on merging this until these changes are made?
Or would you rather that I go ahead and merge now and you make the changes in a separate PR?